snapshot_create.go

```diff
@@ -0,0 +1,179 @@
+func (s *SnapshotCreateService) BodyString(body string) *SnapshotCreateService {
+	s.bodyString = body
+	return s
+}
+
+// buildURL builds the URL for the operation.
+func (s *SnapshotCreateService) buildURL() (string, url.Values, error) {
+	// Build URL
+	path, err := uritemplates.Expand("/_snapshot/{repository}/{snapshot}", map[string]string{
+		"snapshot":   s.snapshot,
+		"repository": s.repository,
+	})
+	if err != nil {
+		return "", url.Values{}, err
+	}
+
+	// Add query string parameters
+	params := url.Values{}
+	if s.pretty {
+		params.Set("pretty", "1")
+	}
+	if s.masterTimeout != "" {
+		params.Set("master_timeout", s.masterTimeout)
+	}
+	if s.waitForCompletion != nil {
+		params.Set("wait_for_completion", fmt.Sprintf("%v", *s.waitForCompletion))
+	}
+	return path, params, nil
+}
+
+// Validate checks if the operation is valid.
+func (s *SnapshotCreateService) Validate() error {
+	var invalid []string
+	if s.repository == "" {
+		invalid = append(invalid, "Repository")
+	}
+	if s.snapshot == "" {
+		invalid = append(invalid, "Snapshot")
+	}
+	if len(invalid) > 0 {
+		return fmt.Errorf("missing required fields: %v", invalid)
+	}
+	return nil
+}
+
+// Do executes the operation.
+func (s *SnapshotCreateService) Do(ctx context.Context) (*SnapshotCreateResponse, error) {
+	// Check pre-conditions
+	if err := s.Validate(); err != nil {
+		return nil, err
+	}
+
+	// Get URL for request
+	path, params, err := s.buildURL()
+	if err != nil {
+		return nil, err
+	}
+
+	// Setup HTTP request body
+	var body interface{}
+	if s.bodyJson != nil {
+		body = s.bodyJson
+	} else {
+		body = s.bodyString
+	}
+
+	// Get HTTP response
+	res, err := s.client.PerformRequest(ctx, "PUT", path, params, body)
+	if err != nil {
+		return nil, err
+	}
+
+	// Return operation response
+	ret := new(SnapshotCreateResponse)
+	if err := json.Unmarshal(res.Body, ret); err != nil {
+		return nil, err
+	}
+	return ret, nil
+}
+
+// SnapshotCreateResponse is the response of SnapshotCreateService.Do.
+type SnapshotCreateResponse struct {
+	// Accepted indicates whether the request was accepted by elasticsearch.
+	// It's available when waitForCompletion is false.
+	Accepted *bool `json:"accepted"`
+
+	// Snapshot is available when waitForCompletion is true.
+	Snapshot *struct {
+		Snapshot          string        `json:"snapshot"`
+		UUID              string        `json:"uuid"`
+		VersionID         int           `json:"version_id"`
+		Version           string        `json:"version"`
+		Indices           []string      `json:"indices"`
+		State             string        `json:"state"`
+		StartTime         time.Time     `json:"start_time"`
+		StartTimeInMillis int64         `json:"start_time_in_millis"`
+		EndTime           time.Time     `json:"end_time"`
+		EndTimeInMillis   int64         `json:"end_time_in_millis"`
+		DurationInMillis  int           `json:"duration_in_millis"`
+		Failures          []interface{} `json:"failures"`
+		Shards            struct {
```

Comment:
There is a [`shardsInfo` in `errors.go`](https://github.com/olivere/elastic/blob/1368b91516f39100baf5e23a02ae9f090c0d90e2/errors.go#L128-L133) that you can use here.